File: hs/src/Q800/Q876.hs
module Q800.Q876
(
    middleNode
) where

middleNode :: [a] -> [a]
middleNode arr = plana arr arr
    where
        plana [] s = s
        plana [_] s = s
        plana [_,_] (_:sx) = sx
        plana (_:_:fx) (_:sx) = plana fx sx

File: golang/q800/q876.go
package q800

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

import . "github.com/zoroqi/stupid-self/golang"

func middleNode(head *ListNode) *ListNode {
	fast, slow := head, head
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		slow = slow.Next
	}
	return slow
}
